django/db/models/query.py

```diff
@@ -1618,14 +1622,44 @@ def all(self):
         """
         return self._chain()
 
-    def filter(self, *args, **kwargs):
+    def _class_filter(cls, *args, **kwargs):
+        if invalid_kwargs := PROHIBITED_FILTER_KWARGS.intersection(kwargs):
+            invalid_kwargs_str = ", ".join(f"'{k}'" for k in sorted(invalid_kwargs))
+            raise TypeError(f"The following kwargs are invalid: {invalid_kwargs_str}")
+        initial_filter = Q(*args, **kwargs)
+        # Chain initial filters.
+        if cls._initial_filter is not None:
+            initial_filter = cls._initial_filter & initial_filter
+            base_class = cls.__bases__[0]
+        else:
+            base_class = cls
+
+        initial_filter_id = id(initial_filter)
```

Comment:
We likely want to memoize of `initial_filter.identity` here to allow x-interpreter caching.